sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala

```diff
@@ -396,6 +417,93 @@ private[spark] object HiveUtils extends Logging {
         config = configurations,
         barrierPrefixes = hiveMetastoreBarrierPrefixes,
         sharedPrefixes = hiveMetastoreSharedPrefixes)
+    } else if (hiveMetastoreJars == "path") {
+
+      def addLocalHiveJars(file: File): Seq[URL] = {
+        if (file.getName == "*") {
+          val files = file.getParentFile.listFiles()
+          if (files == null) {
+            logWarning(s"Hive jar path '${file.getPath}' does not exist.")
+            Nil
+          } else {
+            files.filter(_.getName.toLowerCase(Locale.ROOT).endsWith(".jar")).map(_.toURL).toSeq
+          }
+        } else {
+          file.toURL :: Nil
+        }
+      }
+
+      def checkRemoteHiveJars(path: String): Seq[URL] = {
+        try {
+          val hadoopPath = new Path(path)
+          val fs = hadoopPath.getFileSystem(hadoopConf)
+          if (hadoopPath.getName == "*") {
+            val parent = hadoopPath.getParent
+            if (!fs.exists(parent)) {
+              logWarning(s"Hive Jar ${path} does not exist.")
+              Nil
+            } else if (!fs.getFileStatus(parent).isDirectory) {
+              logWarning(s"Hive Jar ${parent} is not a directory.")
+              Nil
+            } else {
+              fs.listStatus(parent).map(_.getPath.toUri.toURL)
+            }
+          } else {
+            if (!fs.exists(hadoopPath)) {
+              logWarning(s"Hive Jar ${path} does not exist.")
+              Nil
+            } else if (fs.getFileStatus(hadoopPath).isDirectory) {
+              logWarning(s"Hive Jar ${path} not allow directory without `*`")
+              Nil
+            } else {
+              // Since tar/tar.gz file we can't know it's final path yet, not support it
+              hadoopPath.toUri.toURL :: Nil
```

Comment:
How are these jars be downloaded eventually? we'll need to call `FileSystem` API for that later right?

Reply:
> How are these jars be downloaded eventually? we'll need to call `FileSystem` API for that later right?

URLClassLoader will handle this, you can see my previous commit, I use `Utils.fetchFile` to download, but now I remove that.

Reply:
Saw the `fetchFile` call in the previous commit. However I don't see how `URLClassLoader` can handle HDFS paths. I don't see any `FileSystem` related stuff in the class.

Reply:
> Saw the `fetchFile` call in the previous commit. However I don't see how `URLClassLoader` can handle HDFS paths. I don't see any `FileSystem` related stuff in the class.

I have same confuse as you before, until I see this error https://github.com/apache/spark/pull/29881#issuecomment-701211241

Reply:
I see. So Hadoop side has a `URLStreamHandlerFactory` impl which provides handlers for the Hadoop-specific URL schemes. Not sure about the S3 error though :-/

Reply:
> I see. So Hadoop side has a `URLStreamHandlerFactory` impl which provides handlers for the Hadoop-specific URL schemes. Not sure about the S3 error though :-/

Yea,  dig into the code, seems s3 find config will call URLClassLoader and not equal make it reload element again and again.